web_poet/fields.py

```diff
@@ -95,22 +96,39 @@ def __get__(self, instance, owner=None):
 
         @staticmethod
         def _process(value):
-            for processor in out:
-                value = processor(value)
+            if out:
+                for processor in out:
+                    value = processor(value)
             return value
 
         def _processed(self, method):
             """Returns a wrapper for method that calls processors on its result"""
-            if not out:
-                return method
             if inspect.iscoroutinefunction(method):
 
                 async def processed(*args, **kwargs):
+                    await args[0]._validate_input()
                     return self._process(await method(*args, **kwargs))
 
             else:
 
                 def processed(*args, **kwargs):
+                    task = args[0]._validate_input()
+                    try:
+                        asyncio.get_running_loop()
+                    except RuntimeError:
+                        asyncio.run(task)
+                    else:
+                        # TODO: Implement
+                        #
+                        # Running an async function within a sync function
+                        # looks complicated
+                        # https://github.com/django/asgiref/blob/36022636d3390c1c1cbb952336a34d652a6bfb26/asgiref/sync.py#L107-L314
+                        #
+                        # I am not sure if there is a different approach that
+                        # we can take, but I can’t help how much easier things
+                        # would be if we requred the field methods and the
+                        # to_item methods to be async.
```

Comment:
This pull request has basically no implementation yet, because the first step I see would be to determine how to best approach this aspect of the implementation.

The current proposal allows implementing an async `validate_input` method, and when implemented, it is called when accessing a field. However, fields can be both sync and async, and when they are sync, calling the async `validate_input` method from them [seems cumbersome](https://github.com/django/asgiref/blob/36022636d3390c1c1cbb952336a34d652a6bfb26/asgiref/sync.py#L107-L314).

I am considering the following choices at the moment:

- Require `validate_input` to be sync, and only support sync fields from it. While not ideal from a user perspective, it does simplify our implementation, and it may encourage to keep validation computationally cheap.
- Going the [cumbersome approach](https://github.com/django/asgiref/blob/36022636d3390c1c1cbb952336a34d652a6bfb26/asgiref/sync.py#L107-L314), i.e. running the (async) `validate_input` method in a separate thread with its own loop when called from a synchronous field method.
- Making `validate_input` only work when all fields are async, and raising an exception when the method is defined and any field method is not async. This could be problematic in inheritance scenarios, e.g. if you inherit from a class that defines a sync field method, you need to redefine it in your subclass.
- Going full async, i.e. stop allowing sync versions of field and to_item methods altogether as part of the web-poet specification. This would be *very* backward-incompatible, but looking at the existing code to deal with sync/async mixes, and to the “cumbersome approach”, the idea is growing on me.

Reply:
I think only supporting sync validate_input is fine. 

It would disallow some use cases, e.g. validating the input in case the RequestUrl is a dependency, and actual response is retrieved using HttpClient. But in such cases there is another place where the validation can happen - the same method / field where an async request is made.

If we figure out that we do want async validate_input support, and implement one of the solutions, then we'll be able to support the sync version easily, by calling ensure_async.